contracts/implementation/InvariantCheck.sol

```diff
@@ -0,0 +1,66 @@
+//SPDX-License-Identifier: CC-BY-NC-ND-4.0
+pragma solidity 0.8.7;
+
+import "../interfaces/IInvariantCheck.sol";
+import "../interfaces/IPoolCommitter.sol";
+import "../interfaces/IPausable.sol";
+import "../interfaces/ILeveragedPool.sol";
+import "../interfaces/IPoolFactory.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/// @title The contract for checking invariants and pausing if necessary
+/// @notice Every time certain functions are called, known invariants are checked and if any do not hold, contracts are paused.
+contract InvariantCheck is IInvariantCheck {
+    IPoolFactory public immutable poolFactory;
+
+    constructor(address _factory) {
+        poolFactory = IPoolFactory(_factory);
+    }
+
+    /**
+     * @notice Checks all invariants, and pauses all contracts if
+     *         any invariant does not hold.
+     * @dev This should be called before onlyUnpaused, in case contracts are paused then pause check must happen after.
+     */
+    function checkInvariants(address poolToCheck) external override {
+        ILeveragedPool pool = ILeveragedPool(poolToCheck);
+        require(poolFactory.isValidPool(poolToCheck), "Pool is invalid");
+        IPoolCommitter poolCommitter = IPoolCommitter(pool.poolCommitter());
+        uint256 poolBalance = IERC20(pool.quoteToken()).balanceOf(poolToCheck);
+        (
+            IPoolCommitter.TotalCommitment memory totalCommitment,
+            IPoolCommitter.TotalCommitment memory nextTotalCommitment
+        ) = poolCommitter.getPendingCommits();
+        uint256 pendingMints;
+        unchecked {
```

Comment:
Obligatory "why `unchecked`?" question

Reply:
Because for the mint amounts to overflow would require all of the settlement tokens that have been committed to overflow uint256, as in more than 2^256 USDC.

Would you say this is a valid context to use `unchecked`?

Reply:
Am I correct in assuming that the short counterparts will balance each other out? Is that what you're referring to?

Reply:
After thinking about it more I'm not completely confident `unchecked` is _theoretically_ safe (although I am certain that it is practically safe unless we choose a settlement token with a stupidly high `decimals()`), so I've gotten rid of it for now.